### PR TITLE
fixing hosts module making it OS and standard compliant

### PR DESCRIPTION
**Pull request by tacf**
### What does this PR do?

This pull request fixes salt Hosts module (and state) to be Standard compliant and work across OSes, fixing a known Windows limitation. For details views issues referenced bellow.

### What issues does this PR fix or reference?
https://github.com/saltstack/salt/issues/49889
https://github.com/saltstack/salt/issues/29592

### Previous Behavior
- More than 264 chars per line ([Standard Complience](https://www.uow.edu.au/~blane/netapp/ontap/nag/networking/concept/c_oc_netw_maintaining_host_file_limits.html))
- More than 8 alias per line ([Windows Limitation](https://superuser.com/questions/932112/is-there-a-maxium-number-of-hostname-aliases-per-line-in-a-windows-hosts-file))
- Alias and Fqdn mixed together (man (5) hosts)

### New Behavior
- Limited to 264 char per line
- Max 8 alias per line
- fqdn appear before aliases

### Tests written?

No

### Commits signed with GPG?

No

### CODE REVIEW

File: salt/modules/hosts.py [hunk 1]
```diff
@@ -34,7 +32,7 @@ def _get_or_create_hostfile():
     if hfn is None:
         hfn = ''
     if not os.path.exists(hfn):
-        with salt.utils.files.fopen(hfn, 'w'):
+        with salt.utils.fopen(hfn, 'w'):
```
Comment (damon-atkins) [inline salt/modules/hosts.py]: Did you based this on recent code?  Strange to be changing `salt.utils.files.fopen` to `salt.utils.fopen`

**Review by damon-atkins — CHANGES_REQUESTED** [review]
Thanks for raising a PR. The limits you have found are they the same on all platforms? Which platforms have you tested the limits on?

File: salt/modules/hosts.py (hunk 1 above)
Comment (tacf) [reply salt/modules/hosts.py]: This had to be done due to the version we're using, so i'm assuming that the one i removed is the newer way of doing it right? I'll gladly change it back.

**Comment by damon-atkins** [comment]
You need to take a copy from the develop branch and update it. OR raise the pull request against the version of salt you updated/changed.

File: salt/modules/hosts.py (hunk 1 above)
Comment (jcrsilva) [reply salt/modules/hosts.py]: @damon-atkins these limits are only present in Windows as far as we can tell, but we did test on multiple versions of Windows Server (2016 and 2012).

Applying these limits to *nix OSs will make the hosts file more readable tho.

I couldn't find any info on this on an RFC per se, but [articles like these litter the internet](https://www.uow.edu.au/~blane/netapp/ontap/nag/networking/concept/c_oc_netw_maintaining_host_file_limits.html) (whithout citing sources).
I did test that at least in Ubuntu Xenial, the 255 char limit is not enforced.

**Comment by cachedout** [comment]
@tacf This is failing a series of tests I'm afraid.

**Comment by thatch45** [comment]
@tacf I like what you are trying to do here and we need to get this fix in. But I suspect that the failing tests are relative to changing some of the default paths here.
If we don't hear back in a few days with changes I will close this out, that would nto mean that we don't want this contribution, just that when a PR gets stale we should likely refrsh the process.
I would also ask that if you are going to do dedicated docs/whitespace cleanup that we put that in a separate PR, that makes it easier to get those trivial PRs in and get the main code that needs to be fixed in more easily. :)

**Comment by dwoz** [comment]
@tacf Can you please resolve the conflicts and update the branch with the lates changes from develop? That should take care of some of the test noise.

**Review by dwoz — COMMENTED** [review]
@tacf Are you able to rebase this to the `master` branch?

**Comment by leeclemens** [comment]
What issue does this PR address? I see two listed with a question mark as to which one to use. My primary concerns are if it addresses the concerns I raised in the PR I was involved in and if sufficient tests will be introduced to ensure consistent behavior (the concerns I raised).

Are we sure we want to introduce this change in behavior:
`Alias and Fqdn mixed together (man (5) hosts)`
to
`fqdn appear before aliases`

Is there an RFC that says FQDN's should appear before aliases? I believe aliases can be FQDN's (please correct me if I'm wrong, but a default install of CentOS 7 has FQDNs as aliases).

Side note: One of the references is to uow.edu.au which states "Maximum number of name servers is 3" in /etc/hosts. I am not aware of nameservers being related to this /etc/hosts change or being declared modernly in /etc/hosts file. Not trying to be pedantic but it also claims "No file size limit", which cannot be true as every file system has a file size limit. It also makes a few other claims I cannot verify from the man page or RFC 952.

File: salt/modules/hosts.py (hunk 1 above)
Comment (leeclemens) [reply salt/modules/hosts.py]: NetApp says [Maximum line size is 1022 characters](https://library.netapp.com/ecmdocs/ECMP1368834/html/GUID-C6F3B6D1-232D-44BB-A76C-3304C19607A3.html). Also [Maximum number of aliases is 34](https://library.netapp.com/ecmdocs/ECMP1368834/html/GUID-C6F3B6D1-232D-44BB-A76C-3304C19607A3.html)

Comment (leeclemens) [reply salt/modules/hosts.py]: NetApp was the 4th google result. I didn't see the link @jcrsilva referenced from uow.edu.au after 5 pages of google search for "hosts file line limit", so "litters the internet" seems a bit dubious.

**Comment by leeclemens** [comment]
> This pull request fixes salt Hosts module (and state) to be Standard compliant and work across OSes, fixing a known Windows limitation

- What is not currently Standard compliant?
- If it needs to be "fixed" to work with a Standard compliant OS's "limitation", how is that Standard compliant?

File: salt/modules/hosts.py (hunk 1 above)
Comment (leeclemens) [reply salt/modules/hosts.py]: And there is an RFC referenced by the /etc/hosts man page, 952.

**Comment by Ch3LL** [comment]
@tacf bump here. Any chance we can write some tests and you can respond to @leeclemens 's concerns

**Comment by sagetherage** [comment]
Reviewed in Grooming 2021-APR-05 considering closing, but reopening issues to later fix.

**Comment by sagetherage** [comment]
@tacf can you please rebase against the master branch? we are no longer taking changes against develop.

Discussed in Grooming 2021-APR-19 I will give this 2 more weeks and if no response, will close.

**Comment by leeclemens** [comment]
@sagetherage I am still genuinely interested to know what standard this change is complying to. It appears only to be a personal page on University of Wollongong, Australia's website with no other references.

**Comment by sagetherage** [comment]
> I am still genuinely interested to know what standard this change is complying to. It appears only to be a personal page on University of Wollongong, Australia's website with no other references.

@leeclemens great point!

**Comment by tacf** [comment]
Hi, unfortunately i cannot follow up on this. The issues was found during real world usage and after some major research (google) we found that among several links that reported the limitations we found. They should be easily reproducible but like a mentioned i can't follow up on this atm (since this most likely will require additional testing on newer windows server versions to make sure it still holds).

**Pull request closed without merging**